go/vt/vttablet/tabletserver/twopc.go

```diff
@@ -475,5 +482,56 @@ func (tpc *TwoPC) read(ctx context.Context, conn *connpool.Conn, pq *sqlparser.P
 	if err != nil {
 		return nil, err
 	}
+	// TODO: check back for limit, 10k is too high already to have unresolved transactions.
 	return conn.Exec(ctx, q, 10000, false)
 }
+
+// UnresolvedTransactions returns the list of unresolved transactions
+func (tpc *TwoPC) UnresolvedTransactions(ctx context.Context, abandonTime time.Time) ([]*querypb.TransactionMetadata, error) {
+	conn, err := tpc.readPool.Get(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Recycle()
+
+	bindVars := map[string]*querypb.BindVariable{
+		"time_created": sqltypes.Int64BindVariable(abandonTime.UnixNano()),
+	}
+	qr, err := tpc.read(ctx, conn.Conn, tpc.readUnresolvedTransactions, bindVars)
+	if err != nil {
+		return nil, err
+	}
+	var txs []*querypb.TransactionMetadata
+	var participants []*querypb.Target
+	var currentDtid string
+	var currentDtidState querypb.TransactionState
+	for _, row := range qr.Rows {
+		dtid := row[0].ToString()
+		if currentDtid != dtid {
+			if currentDtid != "" {
+				txs = append(txs, &querypb.TransactionMetadata{
+					Dtid:         currentDtid,
+					State:        currentDtidState,
+					Participants: participants,
+				})
+			}
+			currentDtid = dtid
+			sID, _ := row[1].ToInt()
+			currentDtidState = querypb.TransactionState(sID)
+			participants = nil
+		}
+		participants = append(participants, &querypb.Target{
```

Comment:
it's not easy to follow this logic. maybe a comment or two would help. should the `participants = append(participants` line be inside an `else`?